scvi/data/anndata/fields/_obsm_field.py

```diff
@@ -92,11 +97,100 @@ def transfer_field(
         **kwargs,
     ) -> None:
         super().transfer_field(setup_dict, adata_target, **kwargs)
-        self.validate_field(adata_target)
         self.register_field(adata_target)
 
     def compute_summary_stats(self, adata: AnnData) -> dict:
         super().compute_summary_stats(adata)
         n_continuous_covariates = len(self.obs_keys)
         stat_name = f"n_{self.registry_key}"
         return {stat_name: n_continuous_covariates}
+
+
+class CategoricalJointObsField(JointObsField):
+    """
+    An AnnDataField for a collection of categorical .obs fields in the AnnData data structure.
+
+    Parameters
+    ----------
+    registry_key
+        Key to register field under in data registry.
+    obs_keys
+        Sequence of keys to combine to form the obsm field.
+    """
+
+    def __init__(self, registry_key: str, obs_keys: Optional[List[str]]) -> None:
+        super().__init__(registry_key, obs_keys)
+        self._mappings_key = registry_key
+
+    def _initialize_mappings_dict(self, adata: AnnData) -> None:
+        adata.uns[_constants._SETUP_DICT_KEY][self._mappings_key] = dict(
+            mappings=dict(), keys=[]
+        )
+
+    def _make_obsm_categorical(
+        self, adata: AnnData, category_dict: Optional[Dict[str, List[str]]] = None
+    ) -> None:
+        assert self.obs_keys == self.get_field(adata).columns.tolist()
+        categories = dict()
+        for key in self.obs_keys:
+            if category_dict is None:
+                categorical_obs = adata.obs[key].astype("category")
+                mapping = categorical_obs.cat.categories.to_numpy(copy=True)
+                categories[key] = mapping
+            else:
+                possible_cats = category_dict[key]
+                categorical_obs = adata.obs[key].astype(
+                    CategoricalDtype(categories=possible_cats)
+                )
+        store_cats = categories if category_dict is None else category_dict
+
+        mappings_dict = adata.uns[_constants._SETUP_DICT_KEY][self._mappings_key]
+        mappings_dict[_constants._JO_CM_MAPPINGS_KEY] = store_cats
+        mappings_dict[_constants._JO_CM_KEYS_KEY] = self.obs_keys
+        n_cats_per_key = []
+        for k in self.obs_keys:
+            n_cats_per_key.append(len(store_cats[k]))
+        mappings_dict[_constants._JO_CM_N_CATS_PER_KEY] = n_cats_per_key
+
+    def register_field(self, adata: AnnData) -> None:
+        super().register_field(adata)
+        self._combine_obs_fields(adata)
+        self._initialize_mappings_dict(adata)
+        self._make_obsm_categorical(adata)
+
+    def transfer_field(
+        self,
+        setup_dict: dict,
+        adata_target: AnnData,
+        **kwargs,
+    ) -> None:
+        super().transfer_field(setup_dict, adata_target, **kwargs)
+
+        if self.is_empty:
+            return
+
+        extend_categories = getattr(kwargs, "extend_categories", False)
```

Comment:
it's because this is called over a list of fields here: https://github.com/YosefLab/scvi-tools/pull/1237/files#diff-614a605196c42e8b5f08a458dc15b37502526431a01c1597c4dbbc44d67d4ab9R102.
As a result, we don't have field specific kwargs (unless you pass in a dictionary which is more cumbersome).

Reply:
I'm still not sure I understand, it should be able to be an arg with default as false in the signature?